src/mlpack/tests/main_tests/gmm_train_test.cpp

```diff
@@ -0,0 +1,377 @@
+/**
+ * @file gmm_train_test.cpp
+ * @author Yashwant Singh
+ *
+ * Test mlpackMain() of gmm_train_main.cpp.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#include<string>
+
+#define BINDING_TYPE BINDING_TYPE_TEST
+static const std::string testName = "GmmTrain";
+
+#include <mlpack/core.hpp>
+#include <mlpack/core/util/mlpack_main.hpp>
+#include "test_helper.hpp"
+#include <mlpack/methods/gmm/gmm_train_main.cpp>
+
+#include <boost/test/unit_test.hpp>
+#include "../test_tools.hpp"
+
+using namespace mlpack;
+
+struct GmmTrainTestFixture
+{
+ public:
+  GmmTrainTestFixture()
+  {
+    // Cache in the options for this program.
+    CLI::RestoreSettings(testName);
+  }
+
+  ~GmmTrainTestFixture()
+  {
+    // Clear the settings.
+    bindings::tests::CleanMemory();
+    CLI::ClearSettings();
+  }
+};
+
+void ResetGmmTrainSetting()
+{
+  CLI::ClearSettings();
+  CLI::RestoreSettings(testName);
+}
+
+
+BOOST_FIXTURE_TEST_SUITE(GmmTrainMainTest, GmmTrainTestFixture);
+
+// To check if the gaussian is positive or not.
+BOOST_AUTO_TEST_CASE(GmmTrainValidGaussianTest)
+{
+  arma::mat inputData(5, 10, arma::fill::randu);
+
+  SetInputParam("input", std::move(inputData));
+  SetInputParam("gaussians", 0); // Invalid
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+  * To check if the number of gaussians in the output model is same as
+  * that of input gaussian parameter or not.
+ **/
+BOOST_AUTO_TEST_CASE(GmmTrainOutputModelGaussianTest)
+{
+  arma::mat inputData(5, 10, arma::fill::randu);
+
+  SetInputParam("input", std::move(inputData));
+  SetInputParam("gaussians", (int) 2);
+  SetInputParam("trials", (int) 2);
+
+  mlpackMain();
+
+  GMM* gmm = CLI::GetParam<GMM*>("output_model");
+  BOOST_REQUIRE_EQUAL(gmm->Gaussians(), (int) 2);
+}
+
+// Max iterations must be positive.
+BOOST_AUTO_TEST_CASE(GmmTrainMaxIterationsTest)
+{
+  arma::mat inputData(5, 10, arma::fill::randu);
+
+  SetInputParam("input", std::move(inputData));
+  SetInputParam("gaussians", (int) 2);
+  SetInputParam("trials", (int) 1);
+  SetInputParam("max_iterations", (int)-1); // Invalid.
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+// Ensure that Trials must be greater than 0.
```

Comment:
We should also test that the number of trials makes a difference.

Reply:
Add the test for trials.